algorithms/mopt/mopt_optimizer.cpp

```diff
@@ -0,0 +1,180 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+#include "fuzzuf/algorithms/mopt/mopt_optimizer.hpp"
+
+#include <random>
+#include <vector>
+
+#include "fuzzuf/mutator/havoc_case.hpp"
+#include "fuzzuf/optimizer/keys.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+namespace fuzzuf::optimizer {
+
+using fuzzuf::mutator::INSERT_AEXTRA;
+using fuzzuf::mutator::INSERT_EXTRA;
+using fuzzuf::mutator::OVERWRITE_WITH_AEXTRA;
+using fuzzuf::mutator::OVERWRITE_WITH_EXTRA;
+
+MOptParticle::MOptParticle() {}
+MOptParticle::~MOptParticle() {}
+
+MOptOptimizer::MOptOptimizer() {
+  fuzzuf::optimizer::Store::GetInstance().InitKey(
+      fuzzuf::optimizer::keys::NewTestcases, (u64)0);
+  fuzzuf::optimizer::Store::GetInstance().InitKey(
+      fuzzuf::optimizer::keys::HavocOperatorFinds);
+  fuzzuf::optimizer::Store::GetInstance().InitKey(
+      fuzzuf::optimizer::keys::SelectedCaseHistogram);
+  fuzzuf::optimizer::Store::GetInstance().InitKey(
+      fuzzuf::optimizer::keys::LastSpliceCycle, (u32)0);
+
+  for (auto& p : swarm) {
+    for (auto& pos : p.position)
+      pos = fuzzuf::utils::random::Random<double>(min_position, max_position);
+    p.velocity.fill(0);
+  }
+
+  idx = 0;
+  time = 0;
+
+  UpdateInertia();
+}
+
+MOptOptimizer::~MOptOptimizer() {}
+
+void MOptOptimizer::SetScore(size_t i, double score) {
+  auto& p = swarm[idx];
+  p.fitness[i] = score;
+}
+
+void MOptOptimizer::UpdateLocalBest() {
+  auto& p = swarm[idx];
+
+  if (unlikely(time == 0)) {
+    for (size_t i = 0; i < p.fitness.size(); i++) {
+      p.best_fitness[i] = p.fitness[i];
+      p.best_position[i] = p.position[i];
+    }
+  }
+
+  for (size_t i = 0; i < p.fitness.size(); i++) {
+    if (p.fitness[i] > p.best_fitness[i]) {
+      p.best_fitness[i] = p.fitness[i];
+      p.best_position[i] = p.position[i];
+    }
+  }
+}
+
+void MOptOptimizer::UpdateGlobalBest() {
+  if (unlikely(time == 0)) {
+    // TODO?
+  }
+
+  std::array<u64, NUM_CASE> havoc_operator_dist;
+  havoc_operator_dist.fill(0);
+
+  for (size_t i = 0; i < havoc_operator_finds[0].size(); i++) {
+    havoc_operator_dist[i] =
+        havoc_operator_finds[0][i] + havoc_operator_finds[1][i];
+  }
+
+  std::discrete_distribution<u32> dist(havoc_operator_dist.begin(),
+                                       havoc_operator_dist.end());
+  std::vector<double> prob = dist.probabilities();
+
+  for (size_t i = 0; i < prob.size(); i++) {
+    best_position[i] = prob[i];
+  }
+}
+
+void MOptOptimizer::UpdateInertia() {
+  w = (W_INIT - W_END) * (G_MAX - g_now) / G_MAX + W_END;
+  ++g_now %= G_MAX;
```

Comment:
I think these values are a kind of hyperparameter and can be transformed into constexpr functions as done for AFL's hyperparameters.

Reply:
is it possible to convert this function to constexpr even it depends on non-const variable (g_now)?

Reply:
No I just meant that, for example, converting `W_INIT` into `constexpr SomeIntTypeIDontKnow GetWInit() { return SOME_CONSTANT_VALUE; }`.

Reply:
is it assumed to be implemented in MOptOption?

Reply:
I suppose so. But I'm doubting this conversion is needed considering that nothing will want to customize either PSO or MOpt

Reply:
so let's leave this one as it is :)